### DISCUSSION TITLE
caplog in Reporting test?

### DISCUSSION THREAD

**Discussion by raysalemi** [pyuvm/pyuvm#17] [Q&A] [answered]
@tudortimi, could you explain this code?

```
@pytest.mark.parametrize('level,msg', [(logging.DEBUG, 'debug'),
                                       (logging.INFO, 'info'),
                                       (logging.ERROR, 'error'),
                                       (logging.CRITICAL, 'critical')])
def test_logging(level, msg, caplog):
    ro = uvm_report_object('ro')
    ro.logger.propagate = True  # workaround for 'caplog'
    caplog.set_level(logging.DEBUG, ro.logger.name)
    ro.logger.log(level, msg)
    assert caplog.record_tuples == [(ro.logger.name, level, msg)]
```

I see that `@pytest.amrk.paramterize()` sets the values for `level` and `msg,` but where did `caplog` come from?

**Comment by tudortimi** [comment, marked as answer]
`caplog` is a `pytest` built-in fixture: https://docs.pytest.org/en/6.2.x/reference.html#caplog

**Reply by tudortimi** [reply]
`pyest` will see that we're referencing it and it will pass us an object for it auto-magically.

**Reply by raysalemi** [reply]
Oh. Magic.

Just what I need . . . 

:)

**Comment by raysalemi** [comment]
BTW. I released 2.1 today, which has your current stat of the RAL.  I didn't mention it because I think it's partially complete, correct?  It would need a more prominent release number anyway.

**Reply by tudortimi** [reply]
Yes, it's still missing a couple of classes, notably adapter and predictor.

**Comment by raysalemi** [comment]
One day I’ll have the time to dig in. I've got to finish my book first.

**Comment by tudortimi** [comment]
I can give you the gist of it:
- register operation: models an abstract register operation on a memory mapped bus, like read/write to a certain address
- adapter: convert from concrete bus transaction (like AHB, etc.) to register operation and vice-versa
- predictor: subscriber that gets bus transactions and uses an adapter and a register map to figure out which register is being accessed. Triggers access on register.
- register: aggregates fields
- field: stores value, contains logic for accesses (i.e. read/write, read-only, write-only, write-once, etc.) which updates value
- register map: models where registers live in the address space

The user needs to provide an adapter for their bus. The user has to create a register model by creating fields, adding them to registers and mapping the registers to a map. The user needs to instantiate a predictor, supply it with the adapter and with the map, and connect it to their monitor.

**Reply by tudortimi** [reply]
We have some code for the last 3 (register, field, map). We need to add a bit more code for the first 3 to have something we can put together in an example.
